rcljava/cmake/add_source_jar.cmake

```
@@ -0,0 +1,194 @@
+function(add_source_jar _TARGET_NAME)
+
+    cmake_parse_arguments(_add_source_jar
+      ""
+      "VERSION;OUTPUT_DIR;OUTPUT_NAME;MANIFEST"
+      "SOURCES"
+      ${ARGN}
+    )
+
+    # In CMake < 2.8.12, add_jar used variables which were set prior to calling
+    # add_jar for customizing the behavior of add_jar. In order to be backwards
+    # compatible, check if any of those variables are set, and use them to
+    # initialize values of the named arguments. (Giving the corresponding named
+    # argument will override the value set here.)
+    #
+    # New features should use named arguments only.
+    if(NOT DEFINED _add_source_jar_VERSION AND DEFINED CMAKE_JAVA_TARGET_VERSION)
+        set(_add_source_jar_VERSION "${CMAKE_JAVA_TARGET_VERSION}")
+    endif()
+    if(NOT DEFINED _add_source_jar_OUTPUT_DIR AND DEFINED CMAKE_JAVA_TARGET_OUTPUT_DIR)
+        set(_add_source_jar_OUTPUT_DIR "${CMAKE_JAVA_TARGET_OUTPUT_DIR}")
+    endif()
+    if(NOT DEFINED _add_source_jar_OUTPUT_NAME AND DEFINED CMAKE_JAVA_TARGET_OUTPUT_NAME)
+        set(_add_source_jar_OUTPUT_NAME "${CMAKE_JAVA_TARGET_OUTPUT_NAME}")
+        # reset
+        set(CMAKE_JAVA_TARGET_OUTPUT_NAME)
+    endif()
+
+#------------------------------------------------------
+    set(_JAVA_SOURCE_FILES ${_add_source_jar_SOURCES} ${_add_source_jar_UNPARSED_ARGUMENTS})
+
+    if (NOT DEFINED _add_source_jar_OUTPUT_DIR)
+        set(_add_source_jar_OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR})
+    else()
+        get_filename_component(_add_source_jar_OUTPUT_DIR ${_add_source_jar_OUTPUT_DIR} ABSOLUTE)
+    endif()
+    # ensure output directory exists
+    file (MAKE_DIRECTORY "${_add_source_jar_OUTPUT_DIR}")
+
+    if (_add_source_jar_MANIFEST)
+        set(_MANIFEST_OPTION m)
+        get_filename_component (_MANIFEST_VALUE "${_add_source_jar_MANIFEST}" ABSOLUTE)
+    endif ()
```

style nit: For consistency with rest of this project, no spaces before open parenthesis. The same comment applies throughout this file.

```suggestion
    if(NOT DEFINED _add_source_jar_OUTPUT_DIR)
        set(_add_source_jar_OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR})
    else()
        get_filename_component(_add_source_jar_OUTPUT_DIR ${_add_source_jar_OUTPUT_DIR} ABSOLUTE)
    endif()
    # ensure output directory exists
    file(MAKE_DIRECTORY "${_add_source_jar_OUTPUT_DIR}")

    if(_add_source_jar_MANIFEST)
        set(_MANIFEST_OPTION m)
        get_filename_component(_MANIFEST_VALUE "${_add_source_jar_MANIFEST}" ABSOLUTE)
    endif()
```